### PR TITLE
PHP8 Support

### LINKED ISSUE #26: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [ ] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [ ] Modify `composer.json` to drop support for PHP less than 7.3
- [ ] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [ ] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [ ] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [ ] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by javabudd**
References https://github.com/laminas-api-tools/api-tools-hal/issues/26 and targets the 1.7.x branch.

Issues with required libraries:
- laminas-api-tools/api-tools-api-problem (https://github.com/laminas-api-tools/api-tools-api-problem/issues/12)
- laminas/laminas-filter (https://github.com/laminas/laminas-filter/issues/19)
- laminas/laminas-http (https://github.com/laminas/laminas-http/issues/44)
- laminas/laminas-mvc (https://github.com/laminas/laminas-mvc/issues/66)
- laminas/laminas-paginator (https://github.com/laminas/laminas-paginator/issues/16)
- laminas/laminas-view awaiting 2.12.x release

### CODE REVIEW

**Comment by javabudd** [comment]
The failing build on the `PHP 8.0 DEPS=lowest` matrix is due to deprecation warnings being triggered by \Laminas\Mvc\Controller\LazyControllerAbstractFactory::resolveParameter (getClass()).

I'm not sure why deprecation warnings are making the builds fail, adding `@$controller->getEvent()->willReturn($event);` to \LaminasTest\ApiTools\Hal\Plugin\HalTest::setUp:140 fixes the builds, but I don't like it. 

Thoughts?

File: phpunit.xml.dist [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
-    </testsuites>
-    <filter>
-        <whitelist processUncoveredFilesFromWhitelist="true">
-            <directory suffix=".php">src</directory>
-            <exclude>
-                <file>./src/Extractor/EntityExtractorHydratorV3.php</file>
-            </exclude>
-        </whitelist>
-    </filter>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" bootstrap="./vendor/autoload.php" colors="true" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd">
```
Comment (froschdesign) [inline phpunit.xml.dist]: Please use the local XML schema definition file otherwise this line must be updated with every new version.

Comment (javabudd) [reply phpunit.xml.dist]: this file was generated through phpunit using `--migrate-configuration`, which local XML schema are you referring to?

Comment (froschdesign) [reply phpunit.xml.dist]: `vendor/phpunit/phpunit/phpunit.xsd`

File: src/Extractor/LinkCollectionExtractor.php [hunk 2]
```diff
@@ -19,34 +19,25 @@ class LinkCollectionExtractor implements LinkCollectionExtractorInterface
...
     public function __construct(LinkExtractorInterface $linkExtractor)
     {
         $this->setLinkExtractor($linkExtractor);
     }
 
-    /**
-     * @return LinkExtractorInterface
-     */
-    public function getLinkExtractor()
+    public function getLinkExtractor(): LinkExtractorInterface
```
Comment (froschdesign) [inline src/Extractor/LinkCollectionExtractor.php]: Please revert all these changes because if the API for this class is changed we have a BC break here. 
And the topic of this pull request is the support of PHP 8, everything else must be done in a separate pull request.

Thanks in advance!

File: src/Extractor/LinkCollectionExtractor.php [hunk 3]
```diff
@@ -19,34 +19,25 @@ class LinkCollectionExtractor implements LinkCollectionExtractorInterface
...
     {
         return $this->linkExtractor;
     }
 
-    /**
-     * @param LinkExtractorInterface $linkExtractor
-     */
-    public function setLinkExtractor(LinkExtractorInterface $linkExtractor)
+    public function setLinkExtractor(LinkExtractorInterface $linkExtractor): void
```
Comment (froschdesign) [inline src/Extractor/LinkCollectionExtractor.php]: The same here.

File: src/Extractor/LinkCollectionExtractor.php [hunk 4]
```diff
@@ -19,34 +19,25 @@ class LinkCollectionExtractor implements LinkCollectionExtractorInterface
...
     {
         $this->linkExtractor = $linkExtractor;
     }
 
     /**
      * @inheritDoc
      */
-    public function extract(LinkCollection $collection)
+    public function extract(LinkCollection $collection): array
```
Comment (froschdesign) [inline src/Extractor/LinkCollectionExtractor.php]: The same here.

File: src/Extractor/LinkExtractor.php [hunk 5]
```diff
@@ -30,33 +27,33 @@ public function __construct(LinkUrlBuilder $linkUrlBuilder)
     /**
      * @inheritDoc
      */
-    public function extract(Link $object)
+    public function extract(Link $link): array
```
Comment (froschdesign) [inline src/Extractor/LinkExtractor.php]: The same here.

File: src/Factory/HalConfigFactory.php [hunk 6]
```diff
@@ -9,14 +9,11 @@
...
 
-class HalConfigFactory
+class HalConfigFactory implements FactoryInterface
 {
-    /**
-     * @param ContainerInterface $container
-     * @return array|\ArrayAccess
-     */
-    public function __invoke(ContainerInterface $container)
+    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
```
Comment (froschdesign) [inline src/Factory/HalConfigFactory.php]: Please also revert all these changes because the topic is still the support of PHP 8 and these interfaces for factories are not needed.

File: src/Factory/HalControllerPluginFactory.php [hunk 7]
```diff
@@ -11,18 +11,12 @@
...
 class HalControllerPluginFactory implements FactoryInterface
 {
-    /**
-     * @param ContainerInterface $container
-     * @param string $requestedName
-     * @param null|array $options
-     * @return Hal
-     */
-    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
+    public function __invoke(ContainerInterface $container, $requestedName, array $options = null): Hal
```
Comment (froschdesign) [inline src/Factory/HalControllerPluginFactory.php]: Please also revert this because `$requestedName` is no longer documented.

**Comment by froschdesign** [comment]
@javabudd 
Thanks for contribution! 👍 

I added some comments. Please undo any unrelated changes and remember that we cannot introduce changes to the API of classes as this would end in a BC break. Thanks in advance!

**Comment by weierophinney** [comment]
@javabudd When you're ready for us to review to merge, don't forget to change the status from "Draft" to "Ready for review"! :smile:

**Comment by javabudd** [comment]
Bumpin for review!
